akka-http-core/src/main/scala/akka/http/impl/engine/ws/Websocket.scala

```diff
@@ -212,13 +131,85 @@ private[http] object Websocket {
       tick ~> merge.in2
 
       BidiShape(
-        routePreparation.inlet,
+        split.in,
         messagePreparation.outlet,
         messageRendering.inlet,
         merge.out)
     }.named("ws-message-api")
   }
 
+  private object bypassRouter extends GraphStage[FanOutShape2[Output, BypassEvent, MessagePart]] {
+    private val in = Inlet[Output]("in")
+    private val bypass = Outlet[BypassEvent]("bypass-out")
+    private val user = Outlet[MessagePart]("message-out")
+
+    val shape = new FanOutShape2(in, bypass, user)
+
+    def createLogic = new GraphStageLogic(shape) {
+
+      setHandler(in, new InHandler {
+        override def onPush(): Unit = {
+          grab(in) match {
+            case b: BypassEvent with MessagePart ⇒ emit(bypass, b, () ⇒ emit(user, b, () ⇒ pull(in)))
+            case b: BypassEvent                  ⇒ emit(bypass, b, () ⇒ pull(in))
+            case m: MessagePart                  ⇒ emit(user, m, () ⇒ pull(in))
```

Comment:
Obviously it generates garbage, whether significant remains to be measured. But the real question is: can it be avoided? Because equally obviously the continuation may need to be suspended, meaning that the creation of an Object is unavoidable (short of building in a crystal ball).